### PR TITLE
Add upgrade-go-version prow  job and patch prow image versions

### PR DESCRIPTION
**Pull request by michaelhtm**
Description of changes:
This PR adds a new periodic Prow job `upgrade-go-version` that checks the latest Go version available in the ECR public gallery and compares it with the version used in our repository (`build_config.yaml`). If a newer Go version is available, it:

1. Updates the `build_config.yaml` file with the latest Go version.
2. Bumps the patch versions of all Prow job images in `images_config.yaml`.
3. Commits the changes to a new branch.
4. Raises this PR against the `image-config-update` branch.

This automation helps us keep our Go version up-to-date and ensures that Prow job images are rebuilt whenever there is a new Go version, without requiring manual intervention. 

The next prow job that will be pushed will read the PR pushed by upgrade-go-version and automatically build and push prow images to ECR. 

In the future we plan to include more build versions that we want to periodically query, and upgrade if needed, to ensure our prow jobs dependencies are always up-to-date, without requiring any human intervention.

TODO for this PR:
- [x] Add upgrade-go-version to images_config.yaml
- [x] Create Dockerfile.upgrade-go-version
- [x] Add command to periodic jinja 
- [ ] Use go_version in build_config.yaml in existing Dockerfiles

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prow/jobs/tools/cmd/command/queryecrpublic/client.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
+package queryecrpublic
```
[the remark this answers is not in the document]
Comment (a-hilaly) [reply prow/jobs/tools/cmd/command/queryecrpublic/client.go:1]: i'm thinking about moving this packge to pkg

File: prow/jobs/tools/cmd/command/root.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	appName      = "ack-build-tools"
+	appShortDesc = "prow-patcher - patch prow images, build, and release"
+	appLongDesc  = "prow-patcher - A tool to patch prow jobs when there is a change to test infra, or when there's a new go version pushed to ECR"
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/root.go:12]: Looks like these are outdated

File: prow/jobs/tools/cmd/command/query_golang_ecr_helper.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	sort.Sort(semver.Collection(versions))
+
+	return versions[len(versions)-1].String(), nil
+}
+
+// func readCurrentBuildGoVersion(filepath string) (string, error) {
+// 	return "", nil
+// }
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/query_golang_ecr_helper.go:47]: forgotten function :)

File: prow/jobs/tools/cmd/command/query_golang_ecr_helper.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		v, err := semver.Parse(tag)
+		if err != nil {
+			return "", fmt.Errorf("error while parsing version %s: %s", tag, err)
+		}
+		versions = append(versions, v)
+	}
+
+	sort.Sort(semver.Collection(versions))
+
+	return versions[len(versions)-1].String(), nil
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/query_golang_ecr_helper.go:42]: what if versions is empty?

File: prow/jobs/tools/cmd/command/query_golang_ecr_helper.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	"fmt"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/aquasecurity/go-version/pkg/semver"
+	"github.com/aws-controllers-k8s/test-infra/command/queryecrpublic"
+	"gopkg.in/yaml.v3"
+)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/query_golang_ecr_helper.go:13]: Need to be reordered using standards mentioned in https://go.dev/wiki/CodeReviewComments#imports

File: prow/jobs/tools/cmd/command/query_go_ecr.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		&optBuildConfigPath, "build-config-bath", "build_config.yaml", "path to build_config.yaml, where all the build versions are stored",
+	)
+	queryCmd.PersistentFlags().StringVar(
+		&optGoEcrRepository, "golang-ecr-repository", "docker/library/golang", "ecr gallery repository for golang",
+	)
+
+	rootCmd.AddCommand(queryCmd)
+}
+
+func startQuery(cmd *cobra.Command, args []string) error {
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/query_go_ecr.go:32]: can we add a comment on top of this function explaining what is does?

**Review by a-hilaly — COMMENTED** [review]
@michaelhtm Great work on this Michael!! We're very close to merging it. I left a few comments and questions below

File: prow/jobs/tools/cmd/command/root.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	)
+	rootCmd.PersistentFlags().StringVar(
+		&prRepoOwner, "merge-repo-owner", "", "Name of the owner (user or org) of the repo to create the PR against. If not specified, the value of the `-source-owner` flag will be used.",
+	)
+	rootCmd.PersistentFlags().StringVar(
+		&prDescription, "pr-text", "", "Text to put in the description of the pull request.",
+	)
+	rootCmd.PersistentFlags().StringVar(
+		&prSubject, "pr-title", "", "Title of the pull request. If not specified, no pull request will be created.",
+	)
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/root.go:73]: Can't these be generate by the job?

File: prow/jobs/tools/cmd/command/root.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		&commitMessage, "commit-message", "", "Content of the commit message.",
+	)
+	rootCmd.PersistentFlags().StringVar(
+		&commitBranch, "commit-branch", "", "Name of branch to create the commit in. If it does not already exists, it will be created using the `base-branch` parameter",
+	)
+	rootCmd.PersistentFlags().StringVar(
+		&baseBranch, "base-branch", "main", "Name of branch to create the `commit-branch` from.",
+	)
+	rootCmd.PersistentFlags().StringVar(
+		&prBranch, "merge-branch", "image-config-update", "Name of branch to create the PR against (the one you want to merge your branch in via the PR).",
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/root.go:63]: ditto

File: cd/auto-generate/auto-generate-controllers.sh [hunk 9]
```diff
@@ -247,7 +247,7 @@ for CONTROLLER_NAME in $CONTROLLER_NAMES; do
     # sync local branch with the origin, if there is a diff the gh pr command
     # prompts for user input
     git pull --rebase >/dev/null
-
+x
```
Comment (a-hilaly) [inline cd/auto-generate/auto-generate-controllers.sh:250]: Typo i guess? :)

File: prow/jobs/jinja/periodics/upgrade-go-version.jinja2 [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  annotations:
+    description: Querys go version in ECR and compare it with versuib in repository. Raises a PR with updated GO_VERSION and bumped prow image versions
+    karpenter.sh/do-not-evict: "true"
+  labels:
+    preset-github-secrets: "true"
+  agent: kubernetes
+  spec:
+    serviceAccountName: periodic-service-account
+    containers:
+      - image: {{ image_context.images["query-go-version"] }}
```
Comment (a-hilaly) [inline prow/jobs/jinja/periodics/upgrade-go-version.jinja2:13]:
```suggestion
      - image: {{ image_context.images["upgrade-go-version"] }}
```

File: prow/jobs/tools/cmd/command/github.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	opts := github.CreateCommitOptions{}
+	newCommit, _, err := client.Git.CreateCommit(ctx, sourceOwner, sourceRepo, commit, &opts)
+	if err != nil {
+		return err
+	}
+
+	// Attach the commit to the master branch.
+	ref.Object.SHA = newCommit.SHA
+	_, _, err = client.Git.UpdateRef(ctx, sourceOwner, sourceRepo, ref, false)
+	return err
```
[the remark this answers is not in the document]
Comment (michaelhtm) [reply prow/jobs/tools/cmd/command/github.go:109]: it is

File: prow/jobs/tools/cmd/command/upgrade_go_version.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/spf13/cobra"
+)
+
+// TODO: need to add more flags to handle making a pull request
+var (
+	OptBuildConfigPath string
+	OptGoEcrRepository string
+)
+
+var queryCmd = &cobra.Command{
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/upgrade_go_version.go:29]:
```suggestion
var upgradeGoVersionCmd = &cobra.Command{
```

File: prow/jobs/tools/cmd/command/upgrade_go_version_helper.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	"fmt"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/aquasecurity/go-version/pkg/semver"
+	queryecrpublic "github.com/aws-controllers-k8s/test-infra/command/ecrpublic"
```
Comment (a-hilaly) [inline prow/jobs/tools/cmd/command/upgrade_go_version_helper.go:26]:
```suggestion
	ecrpublic "github.com/aws-controllers-k8s/test-infra/command/ecrpublic"
```

**Review by a-hilaly — APPROVED** [review]
Excellent work, Michael!! I'm excited to see this in action
/lgtm

**Pull request merged**